Keas.Mvc/Services/TeamsManager.cs

```diff
@@ -0,0 +1,192 @@
+            }
+            else
+            {
+                try
+                {
+                    roleContainer = JsonConvert.DeserializeObject<RoleContainer>(sessionResult);
+                }
+                catch (Exception e)
+                {
+                    Log.Error(e.Message);
+                    roleContainer = new RoleContainer();
+                }
+
+                if (roleContainer.UserId != userId)
+                {
+                    _contextAccessor.HttpContext.Session.Remove(RolesSessionKey);
+                    roleContainer = new RoleContainer();
+                    roleContainer.UserId = userId;
+                }
+            }
+
+            return roleContainer;
+        }
+
+        private async Task<RoleContainer> GetSystemRoles(RoleContainer roleContainer)
+        {
+            var userId = _contextAccessor.HttpContext.User.Identity.Name;
+            var systemRole = roleContainer.SystemRoles;
+            if (systemRole == null)
+            {
+                systemRole = new Team
+                {
+                    TeamName = "System",
+                    TeamRoles = await _dbContext.SystemPermissions
+                        .Where(a => a.UserId == userId).Select(a => a.Role.Name).Distinct().ToArrayAsync(),
+                };
+                roleContainer.SystemRoles = systemRole;
+                _contextAccessor.HttpContext.Session.SetString(RolesSessionKey, JsonConvert.SerializeObject(roleContainer));
+            }
+            return roleContainer;
+        }
+
+        private async Task<RoleContainer> GetTeams(string slug, RoleContainer roleContainer)
+        {
+            var userId = _contextAccessor.HttpContext.User.Identity.Name;
+            var team = roleContainer.Teams.FirstOrDefault(a => a.TeamName == slug);
+
+            if (team == null)
+            {
+                team = new Team
+                {
+                    TeamName = slug,
+                    TeamRoles = await _dbContext.TeamPermissions
+                        .Where(a => a.Team.Slug == slug && a.UserId == userId)
+                        .Select(a => a.Role.Name).ToArrayAsync(),
+                };
+                roleContainer.Teams.Add(team);
+
+                _contextAccessor.HttpContext.Session.SetString(RolesSessionKey, JsonConvert.SerializeObject(roleContainer));
+            }
+
+            return roleContainer;
+        }
+
+
+        public class RoleContainer
+        {
+            public RoleContainer()
+            {
+                Teams = new List<Team>();
+            }
+            public string UserId { get; set; }
+            public List<Team> Teams { get; set; }
+
+            public Team SystemRoles { get; set; }
+            public Dictionary<string, string> TeamDictionary { get; set; }
+        }
+
+        public class Team
+        {
+            public string TeamName { get; set; }
+            public string[] TeamRoles { get; set; }

```

Comment:
Generally I don't think it's necessary to add the classname to the properties of that class, like I'd rather you have `roles.team.name` than `roles.team.teamName`.